[WIP] deprecate view registry

chancancode opened the pull request:
This fixes an issue where event dispatching stops working if `id` is passed with splattributes, because `elementId` will diverge from the actual `id`, causing the element to not be found during event dispatching.

~~This patch does not fix the `elementId` divergent problem, but changes the framework code to not rely on `elementId` and use a weak map instead. I have ideas for fixing the `elementId` issue, and making it optional, but I'll probably submit it separately.~~ (It's fixed.)

It's probably best to review this commit-by-commit.

This also removes the view registry private API. The theory is that there are very few addons that uses it, and we can transition them to using the new `Ember.ViewUtils.getElementView` instead.

rwjblue commented on packages/@ember/-internals/views/lib/system/utils.js:
```diff
@@ -60,23 +60,30 @@ export function getViewId(view) {
+}
+
+// Not needed for GC, but for correctness
+export function clearViewElement(view) {
```
Shouldn't this also remove from `ELEMENT_VIEW`? Seems _slightly_ odd to have this mismatch vs setViewElement

rwjblue reviewed (commented):
Generally 👍, only had one small question

chancancode committed:
[CLEANUP] Remove ViewRegistry

The framework does not need it anymore.

In most cases, `Ember.ViewUtils.getElementView` can be used as a
replacement to get the classic component (`Ember.Component`) that
corresponds to a DOM element (if any).

chancancode committed:
[BUGFIX canary] Ensure `elementId` matches reality

Since the `id` attribute can be passed through splattributes, we cannot
assume the generated `elementId` will be used. Instead of eagerly generating
an `elementId` (if not set in `init`), we make it as lazy as possible and
read from DOM after flushing the attributes. This ensures the value reflects
what was eventually set. If, however, an `elementId` is explicitly set during
`init`, we want to ensure they match up when `elementId` is accessed again.

Note that with this commit, the framework no longer rely on `elementId` for
any purpose. We may want to deprecate _reading_ from `elementId`, making it
a "write-only" API that only serve to customize the wrapper element.

wycats commented on packages/@ember/-internals/views/lib/system/utils.js:
```diff
@@ -60,23 +60,38 @@ export function getViewId(view) {
+  VIEW_ELEMENT.set(view, element);
+}
+
+// These are not needed for GC, but for correctness
```
Can you write a little more here?

rwjblue reviewed (dismissed):
LGTM

mixonic commented:
Should `-view-registry` go through an intimate API deprecation re: [usages in the wild](https://www.emberobserver.com/code-search?codeQuery=-view-registry&sort=usages&sortAscending=false)?

rwjblue commented:
> Should -view-registry go through an intimate API deprecation re: usages in the wild?

I did review usages on EmberObserver.com and I do not think the limited usage merits intimate API deprecation. The usages in addons that are maintained (releases in the last year or so) are mostly official packages (`@ember/test-helpers`, `ember-resolver` in the MU flagged code, etc), the usages in maintained non-official addons seem to mostly be in custom testing utilities and actually do have a nice API to migrate to.

rwjblue commented:
@chancancode - Should we close, or revive?

chancancode commented:
huh yes, I thought we did the elementId fix already.. did we not?

chancancode commented:
Ah we did in fact land #17818 at least. We can probably just remove this now since the inspector switched to using the debug render tree. (Though I just added a test in #19382 that uses the view registry because I was lazy. 😉)

locks commented:
@chancancode is this still relevant?

kategengler commented:
Is this still relevant? It is labeled `wip`.

kategengler commented:
Closing as stale
